### PR TITLE
feat(query): parallel k way merge sort

### PR DESCRIPTION
**Pull request by forsaken628**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary
#16217

The merge path is not used
* It is more suitable for the outer sort algorithm, for which the entire column is accessible. But it is much more troublesome for stream sorting.
* It is only applicable to cascading merge algorithms, we are k-way merging.

## Tests

- [x] Unit Test
- [ ] Logic Test
- [ ] Benchmark Test
- [ ] No Test - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/16340)

### CODE REVIEW

File: src/query/pipeline/transforms/src/processors/transforms/transform_k_way_merge_sort.rs [hunk 1]
```diff
@@ -0,0 +1,817 @@
...
+                None => block,
+                Some(limit) => {
+                    let n = block.num_rows();
+                    if *limit >= n {
+                        *limit -= n;
+                        block
+                    } else {
+                        let range = 0..*limit;
+                        *limit = 0;
+                        block.slice(range)
```
Comment (sundy-li) [inline src/query/pipeline/transforms/src/processors/transforms/transform_k_way_merge_sort.rs:695]: if limit is reached, we should set `input.finish`

File: src/query/pipeline/transforms/src/processors/transforms/transform_k_way_merge_sort.rs [hunk 2]
```diff
@@ -0,0 +1,817 @@
...
+        input.set_need_data();
+        Ok(PullEvent::Pending)
+    }
+
+    fn push(&mut self) -> bool {
+        let cur_input = self.cur_input.unwrap();
+        let buffer = &mut self.buffer[cur_input];
+        let info = &mut self.info[cur_input];
+        debug_assert_eq!(info.unwrap().task, self.cur_task);
+        if let Some(block) = buffer.pop_front() {
```
Comment (sundy-li) [inline src/query/pipeline/transforms/src/processors/transforms/transform_k_way_merge_sort.rs:684]: block may be very small, we need to concat blocks in buffer with MAX_BLOCK_SIZE

File: src/query/pipeline/transforms/src/processors/transforms/transform_k_way_merge_sort.rs (hunk 1 above)
Comment (forsaken628) [reply src/query/pipeline/transforms/src/processors/transforms/transform_k_way_merge_sort.rs:695]: It's finished outside.
https://github.com/datafuselabs/databend/pull/16340/files/0e98460dceb616fa591b664a498c304c00f0e96f#diff-81ba0a693181282337e4ca7db834936257017e273faf054eaec00bf3b123eb29R327-R331

File: src/query/pipeline/transforms/src/processors/transforms/transform_k_way_merge_sort.rs [hunk 3]
```diff
@@ -0,0 +1,817 @@
...
+            output.push_data(Ok(block));
+            if self.task.is_empty() {
+                self.cur_output = None;
+            }
+            return Ok(Event::NeedConsume);
+        }
+
+        if self.partitioner.is_finished() {
+            self.outputs.iter().for_each(|o| o.finish());
+            self.inputs.iter().for_each(|i| i.finish());
```
Comment (forsaken628) [inline src/query/pipeline/transforms/src/processors/transforms/transform_k_way_merge_sort.rs:329]: input finish

**Comment by forsaken628** [comment]
depend on #16401

File: src/query/pipeline/transforms/src/processors/transforms/transform_k_way_merge_sort.rs (hunk 2 above)
Comment (forsaken628) [reply src/query/pipeline/transforms/src/processors/transforms/transform_k_way_merge_sort.rs:684]: > block may be very small, we need to concat blocks in buffer with MAX_BLOCK_SIZE

depend on https://github.com/datafuselabs/databend/pull/16401

**Comment by forsaken628** [comment]
depend on https://github.com/datafuselabs/databend/issues/16412

File: tests/sqllogictests/suites/base/09_fuse_engine/09_0008_fuse_optimize_table.test [hunk 4]
```diff
@@ -836,7 +836,7 @@ alter table t15 recluster
 query TTIIFFT
 select * exclude(timestamp) from clustering_information('db_09_0008','t15')
 ----
-(abs(a)) linear 3 0 0.0 1.0 {"00001":3}
+(abs(a)) linear 2 0 0.0 1.0 {"00001":2}
```
Comment (forsaken628) [inline tests/sqllogictests/suites/base/09_fuse_engine/09_0008_fuse_optimize_table.test:839]: @zhyass please take a look

**Comment by forsaken628** [comment]
@dosu Tell me if there is something bad about this PR in particular:
* Inconsistent
* Difficult to understand
* Lacks necessary comment
* Poorly named
* Could be simplified
* Pitfalls
* Redundant and repetitive
Preferably with suggested changes.

File: tests/sqllogictests/suites/base/20+_others/20_0013_query_result_cache.test [hunk 5]
```diff
@@ -13,6 +13,9 @@ CREATE TABLE IF NOT EXISTS t1 (a INT not null);
 statement ok
 INSERT INTO t1 VALUES (1), (2), (3);
 
+statement ok
```
Comment (forsaken628) [inline tests/sqllogictests/suites/base/20+_others/20_0013_query_result_cache.test:16]: The size of the result cache after parallel sorting is variable, sometimes 117 and sometimes 125.

File: tests/sqllogictests/suites/base/20+_others/20_0013_query_result_cache.test [hunk 6]
```diff
@@ -13,6 +13,9 @@ CREATE TABLE IF NOT EXISTS t1 (a INT not null);
...
+set enable_parallel_multi_merge_sort = 0;
```
Comment (forsaken628) [inline tests/sqllogictests/suites/base/20+_others/20_0013_query_result_cache.test:17]: The size of the result cache after parallel sorting is variable, sometimes 117 and sometimes 125.

**Review by sundy-li — APPROVED** [review]

File: tests/sqllogictests/suites/base/09_fuse_engine/09_0016_remote_alter_recluster.test [hunk 7]
```diff
@@ -78,7 +78,7 @@ ALTER TABLE t3 RECLUSTER
 query FFT
 select average_overlaps, average_depth, block_depth_histogram from clustering_information('db_09_0016','t3')
 ----
-0.0 1.0 {"00001":2}
+0.0 1.0 {"00001":1}
```
Comment (zhyass) [inline tests/sqllogictests/suites/base/09_fuse_engine/09_0016_remote_alter_recluster.test:81]: Remove `block_depth_historgram`.
```sql
select average_overlaps, average_depth from clustering_information('db_09_0016','t3')
----
0.0 1.0
```

**Pull request merged**